src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/CodeownersGenerator.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.DotNet.Darc.Models.VirtualMonoRepo;
+using Microsoft.DotNet.DarcLib.Helpers;
+using Microsoft.Extensions.Logging;
+
+#nullable enable
+namespace Microsoft.DotNet.DarcLib.VirtualMonoRepo;
+
+public interface ICodeownersGenerator
+{
+    Task UpdateCodeowners(CancellationToken cancellationToken);
+}
+
+public class CodeownersGenerator : ICodeownersGenerator
+{
+    private const string CodeownersHeader = "### CONTENT BELOW IS AUTO-GENERATED AND MANUAL CHANGES WILL BE OVERWRITTEN ###";
+
+    private static readonly IReadOnlyCollection<LocalPath> s_codeownersLocations = new[]
+    {
+        new UnixPath(VmrInfo.CodeownersFileName),
+        new UnixPath(".github/" + VmrInfo.CodeownersFileName),
+        new UnixPath("docs/" + VmrInfo.CodeownersFileName),
+    };
+
+    private readonly IVmrInfo _vmrInfo;
+    private readonly ISourceManifest _sourceManifest;
+    private readonly ILocalGitRepo _localGitClient;
+    private readonly IFileSystem _fileSystem;
+    private readonly ILogger<CodeownersGenerator> _logger;
+
+    public CodeownersGenerator(
+        IVmrInfo vmrInfo,
+        ISourceManifest sourceManifest,
+        ILocalGitRepo localGitClient,
+        IFileSystem fileSystem,
+        ILogger<CodeownersGenerator> logger)
+    {
+        _vmrInfo = vmrInfo;
+        _sourceManifest = sourceManifest;
+        _localGitClient = localGitClient;
+        _fileSystem = fileSystem;
+        _logger = logger;
+    }
+
+    /// <summary>
+    /// Generates the CODEOWNERS file by gathering individual repo CODEOWNERS files.
+    /// </summary>
+    public async Task UpdateCodeowners(CancellationToken cancellationToken)
+    {
+        _logger.LogInformation("Updating {codeowners}...", VmrInfo.CodeownersPath);
+
+        var destPath = _vmrInfo.VmrPath / VmrInfo.CodeownersPath;
+
+        string header = CodeownersHeader;
+        if (_fileSystem.FileExists(destPath))
+        {
+            var content = await _fileSystem.ReadAllTextAsync(destPath);
+
+            int position = content.IndexOf(CodeownersHeader);
+            if (position != -1)
+            {
+                header = content.Substring(0, position + CodeownersHeader.Length);
+            }
+        }
+
+        _fileSystem.CreateDirectory(_fileSystem.GetDirectoryName(destPath)
+            ?? throw new Exception($"Failed to create {VmrInfo.CodeownersFileName} in {destPath}"));
+
+        using (var destStream = _fileSystem.GetFileStream(destPath, FileMode.Create, FileAccess.Write))
+        using (var writer = new StreamWriter(destStream, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false)))
+        {
+            if (!string.IsNullOrEmpty(header))
```

Comment:
Can the header ever be empty?

Reply:
Yes, when the file is empty or there is just not anything above the auto-generated part